src/aws-util/src/s3_uploader.rs

```diff
@@ -225,23 +224,29 @@ impl S3MultiPartUploader {
     pub async fn finish(mut self) -> Result<CompletedUpload, S3MultiPartUploadError> {
         if self.buffer.len() > 0 {
             let remaining = self.buffer.split();
-            self.upload_part_internal(remaining.freeze()).await?;
+            self.upload_part_internal(remaining.freeze())?;
         }
 
         if self.part_count < AWS_S3_MIN_PART_COUNT {
             return Err(S3MultiPartUploadError::AtLeastMinPartNumber);
         }
 
-        let parts: Vec<CompletedPart> = self
-            .etags
-            .iter()
-            .map(|(part_num, etag)| {
-                CompletedPart::builder()
-                    .e_tag(etag)
-                    .part_number(*part_num)
-                    .build()
-            })
-            .collect();
+        let mut parts: Vec<CompletedPart> = Vec::with_capacity(self.upload_handles.len());
+        for handle in self.upload_handles {
```

Comment:
oooh nice, to do this in `finish`!

Though, you may prefer to use https://docs.rs/tokio/latest/tokio/task/join_set/struct.JoinSet.html!

Reply:
Ah nice that looks useful - punting on refactoring it now in the interest of time